### PR TITLE
MVP for listing a user's previous orders

### PR DESCRIPTION
**Pull request by RVRX**
Very basic list of a user's orders. If we want more data order here, its is all fetched for the current User and ready to be referenced
image

### CODE REVIEW

**Comment by RVRX** [comment]
Once again, will clean up commit structure once DB is settled on. Don't wanna squash any commits until the associations are set in stone

File: lib/lunchbot_web/router.ex [hunk 1]
```diff
@@ -34,6 +34,11 @@ defmodule LunchbotWeb.Router do
     resources "/options", OptionsController
   end
 
+  scope "/", LunchbotWeb do
+    pipe_through [:browser, :require_admin_user]
```
Comment (force10269) [inline lib/lunchbot_web/router.ex]: Could we pipe through `:require_authenticated_user` instead if we just want our user to be logged in to view their previous orders

File: lib/lunchbot_web/live/list_orders_live.ex [hunk 2]
```diff
@@ -85,41 +21,40 @@ defmodule LunchbotWeb.ListOrdersLive do
     orders = Repo.all(Order, user_id: user_id)
     orders = Order |> Order.get_orders_by_user(user_id) |> Repo.all()
 
-    # turn the list of orders in a readable format to be given to the heex template
-    orders_heex =
-      Enum.into(orders, [], fn an_order -> transform(an_order) end)
-      |> IO.inspect(label: "List of user's previous orders:")
+    user_id_with_orders =
```
Comment (force10269) [inline lib/lunchbot_web/live/list_orders_live.ex]: Can we rebase and squash so we have the ecto associations commit first and then squash this fixup with the first MVP commit

File: lib/lunchbot_web/router.ex (hunk 1 above)
Comment (RVRX) [reply lib/lunchbot_web/router.ex]: ahh yup good catch!

File: lib/lunchbot_web/live/list_orders_live.ex (hunk 2 above)
Comment (RVRX) [reply lib/lunchbot_web/live/list_orders_live.ex]: yup thats the plan once i clean the commits up

File: lib/lunchbot/lunchbot_data/order_item_options.ex [hunk 3]
```diff
@@ -21,4 +21,13 @@ defmodule Lunchbot.LunchbotData.OrderItemOptions do
   def get_order_item_options_by_order_item_id(query, order_item_id) do
     from(q in query, where: q.order_item_id == ^order_item_id)
   end
+
+  def get_total(%Lunchbot.LunchbotData.OrderItemOptions{} = order_item_option) do
```
Comment (chewnoill) [inline lib/lunchbot/lunchbot_data/order_item_options.ex:20]:
```sql
select sum(item.price)
from order
join items on item.order_id = order.id

select sum(option.price)
from order
 join items on item.order_id = order.id
 join options on option.item_it = item.id
```

you can use aggregate queries to calculate the sum on the database side

File: lib/lunchbot/lunchbot_data/order_item.ex [hunk 4]
```diff
@@ -21,4 +21,13 @@ defmodule Lunchbot.LunchbotData.OrderItem do
   def get_order_items_by_order_id(query, order_id) do
     from(q in query, where: q.order_id == ^order_id)
   end
+
+  def get_total(%Lunchbot.LunchbotData.OrderItem{} = order_item) do
+    total_options_price =
+      Enum.map(order_item.order_item_options, &Lunchbot.LunchbotData.OrderItemOptions.get_total/1)
```
Comment (chewnoill) [inline lib/lunchbot/lunchbot_data/order_item.ex:22]: we need unit tests around edge cases, ie items/options not having prices

File: lib/lunchbot_web/live/user_auth_live.ex [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  import Phoenix.LiveView
+  alias Lunchbot.Accounts
+
+  # Helper for LiveView pages to get the current user.
+  # To use, declare "on_mount {LunchbotWeb.UserAuthLive, :user}"
+  # and reference in HEEX as "@current_user"
+  def on_mount(:user, _params, %{"user_token" => user_token} = _session, socket) do
+    socket =
+      socket
+      |> assign(:current_user, Accounts.get_user_by_session_token(user_token))
```
Comment (chewnoill) [inline lib/lunchbot_web/live/user_auth_live.ex]: nit: you can fit this all on one line without the pipe?

File: lib/lunchbot/lunchbot_data/order.ex [hunk 6]
```diff
@@ -4,8 +4,10 @@ defmodule Lunchbot.LunchbotData.Order do
 
   schema "orders" do
     field :lunch_order_id, :integer
-    field :menu_id, :integer
+    # :menu_id
+    belongs_to :menu, Lunchbot.LunchbotData.Menu
     field :user_id, :integer
+    has_many :order_items, Lunchbot.LunchbotData.OrderItem
```
Comment (cathyzoller) [inline lib/lunchbot/lunchbot_data/order.ex:11]: Since order_items is a join table, we could add a many to many relationship here, so that later order.items could be used.  Something like:
`many_to_many :items, LunchbotData.Item, join_through: "order_items"` then repeat the other direction in the Item schema.

File: lib/lunchbot/lunchbot_data/order_item.ex [hunk 7]
```diff
@@ -3,8 +3,9 @@ defmodule Lunchbot.LunchbotData.OrderItem do
   import Ecto.Changeset
 
   schema "order_items" do
-    field :item_id, :integer
+    belongs_to :item, Lunchbot.LunchbotData.Item
     field :order_id, :integer
+    has_many :order_item_options, Lunchbot.LunchbotData.OrderItemOptions
```
Comment (cathyzoller) [inline lib/lunchbot/lunchbot_data/order_item.ex:8]: Likewise, could add a many_to_many association between order_items and options in the order_items schema
`many_to_many :options, LunchbotData.Option, join_through: "order_item_options"`
Hmmm... then we'd still need to preload like this anyway to get both the ordered item info & the ordered item options.
`Repo.get!(Order, id) |> Repo.preload([items: [order_items: [:options]]])`
Saves a bit of preload work anyway.

File: lib/lunchbot/lunchbot_data/order_item_options.ex [hunk 8]
```diff
@@ -3,7 +3,8 @@ defmodule Lunchbot.LunchbotData.OrderItemOptions do
   import Ecto.Changeset
 
   schema "order_item_options" do
-    field :option_id, :integer
+    # option_id
+    belongs_to :option, Lunchbot.LunchbotData.Options
```
Comment (cathyzoller) [inline lib/lunchbot/lunchbot_data/order_item_options.ex:7]: This also belongs to :order_item, but perhaps is not added since you're not yet using it?

File: lib/lunchbot/lunchbot_data/order.ex [hunk 9]
```diff
@@ -23,4 +24,21 @@ defmodule Lunchbot.LunchbotData.Order do
...
+  def get_order_total_price(order_id) do
+    order =
+      Lunchbot.Repo.get_by(Lunchbot.LunchbotData.Order, id: order_id)
+      |> Repo.preload(order_items: [:item, order_item_options: [:option]])
+
+    get_total(order)
+  end
+
+  def get_total(%Lunchbot.LunchbotData.Order{} = order) do
+    total_items_price =
```
Comment (cathyzoller) [inline lib/lunchbot/lunchbot_data/order.ex]: In elixir, the last line of code is the return.  In this case the variable isn't needed.  You could write instead:
```
    def get_total(order) do
           Enum.map(order.order_items, &Lunchbot.LunchbotData.OrderItem.get_total/1)
           |> Enum.sum()
     end
```

File: lib/lunchbot_web/live/user_auth_live.ex [hunk 10]
```diff
@@ -0,0 +1,19 @@
+defmodule LunchbotWeb.UserAuthLive do
+  import Phoenix.LiveView
+  alias Lunchbot.Accounts
+
+  # Helper for LiveView pages to get the current user.
```
Comment (cathyzoller) [inline lib/lunchbot_web/live/user_auth_live.ex:5]: Nice.

File: lib/lunchbot_web/live/user_auth_live.ex (hunk 5 above)
Comment (cathyzoller) [reply lib/lunchbot_web/live/user_auth_live.ex]: eg: `socket = assign(socket, :current_user...)` will match the new value on right to the var socket.

**Review by cathyzoller — COMMENTED** [review]
It may (or may not) be useful to use the many_to_many relationships represented by the schema.  Something to think about anyway, before approving.  Perhaps we could discuss.

File: lib/lunchbot/lunchbot_data/order_item_options.ex (hunk 8 above)
Comment (RVRX) [reply lib/lunchbot/lunchbot_data/order_item_options.ex:7]: I think I'm going to hold off on adding the associations on everything for now. @force10269 is adding the ones he needs in his PR and I'm adding only what I need here - when we merge them hopefully it covers most the bases.

File: lib/lunchbot/lunchbot_data/order_item_options.ex (hunk 3 above)
Comment (RVRX) [reply lib/lunchbot/lunchbot_data/order_item_options.ex:20]: After some deliberation with @cbarber + @isometriks we decided that since the result of the current query is also used to fill some other fields on the page, we will be keeping it as is. Would not make sense to use one query to calculate price, then query the same tables again for more data.

File: lib/lunchbot/lunchbot_data/order_item.ex (hunk 7 above)
Comment (RVRX) [reply lib/lunchbot/lunchbot_data/order_item.ex:8]: If I define `many_to_many :items, Lunchbot.LunchbotData.Item, join_through: "order_items"` in the `Orders` schema + the `many_to_many` between order and items, then I don't seem to be able to get to `order_items` -> `order_item_options`/options as Lunchbot.LunchbotData.Order will no longer have an association to :order_items

File: lib/lunchbot/lunchbot_data/item.ex [hunk 11]
```diff
@@ -10,6 +10,7 @@ defmodule Lunchbot.LunchbotData.Item do
     field :item_image, :string
     field :name, :string
     field :price, :integer
+    has_many :order_item, Lunchbot.LunchbotData.OrderItem
```
Comment (cathyzoller) [inline lib/lunchbot/lunchbot_data/item.ex]: :order_item is plural on the has_many side of the relation, and singular on the belongs_to side

File: test/lunchbot/order_test.exs [hunk 12]
```diff
@@ -0,0 +1,129 @@
+defmodule Lunchbot.OrderTest do
```
Comment (cathyzoller) [inline test/lunchbot/order_test.exs:1]: Nice tests

File: lib/lunchbot_web/live/list_orders_live.ex [hunk 13]
```diff
@@ -0,0 +1,50 @@
+defmodule LunchbotWeb.ListOrdersLive do
```
Comment (cathyzoller) [inline lib/lunchbot_web/live/list_orders_live.ex]: The name here could cause confusion with the admin list of orders for the daily lunch order.  How about calling it `PreviousOrdersLive` or even `ListPreviousOrdersLive`

**Review by cathyzoller — CHANGES_REQUESTED** [review]
Just need to pluralize :order_item on line 13 of item.ex. Name change comment is more of a nit.

File: lib/lunchbot/lunchbot_data/item.ex (hunk 11 above)
Comment (RVRX) [reply lib/lunchbot/lunchbot_data/item.ex]: Good catch, ty

**Review by cathyzoller — APPROVED** [review]
Great work @RVRX
